Cryptic Error Message when module-level __getattr__ fails to raise AttributeError

I just ran into an error that was very difficult to figure out. This is not a bug in pytest perse, but it is an area that might benefit from having a better error message. 

In Python 3.7+ you are now allowed to define a custom `__getattr__` at the module level. I find this incredibly useful giving users deprecation errors when they try to use something that was recently moved or removed. However, I made a mistake in my last one, and instead of raising an AttributeError in the fallback case, I forgot and let the default None return.

This actually doesn't hurt anything in normal usage, because only attributes that exist are ever accessed. However, it seems like pytest attempts to do a getattr at the module level in `normalize_mark_list` defined in `_pytest.mark.structures`. Thus if I define the file: 

```python
#pytest_getattr_mwe.py


def __getattr__(key):
    return None
    # raise AttributeError(key)


def test_1():
    return 1
```

and run `pytest pytest_getattr_mwe.py` I get:

```
(py38) joncrall@Ooo:~/misc/tests/python$ pytest pytest_getattr_mwe.py 
=============================================================================== test session starts ================================================================================
platform linux -- Python 3.8.2, pytest-6.2.1, py-1.9.0, pluggy-0.13.1
rootdir: /home/joncrall/misc/tests/python
plugins: celery-4.4.7, timeout-1.4.1, notebook-0.6.0, hydra-core-1.0.5, lazy-fixture-0.6.3, cov-2.8.1, xdoctest-0.15.2
collected 0 items / 1 error                                                                                                                                                        

====================================================================================== ERRORS ======================================================================================
______________________________________________________________________ ERROR collecting pytest_getattr_mwe.py ______________________________________________________________________
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/runner.py:311: in from_call
    result: Optional[TResult] = func()
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/runner.py:341: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/python.py:503: in collect
    self._inject_setup_module_fixture()
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/python.py:516: in _inject_setup_module_fixture
    self.obj, ("setUpModule", "setup_module")
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/python.py:295: in obj
    self.own_markers.extend(get_unpacked_marks(self.obj))
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/mark/structures.py:353: in get_unpacked_marks
    return normalize_mark_list(mark_list)
../../../.local/conda/envs/py38/lib/python3.8/site-packages/_pytest/mark/structures.py:367: in normalize_mark_list
    raise TypeError(f"got {mark!r} instead of Mark")
E   TypeError: got None instead of Mark
```

Perhaps it would be possible to improve this error message? For instance if `mark is None` maybe hint that `__getattr__` may have been ill-defined? (I'm not sure if this can happen in other cases). 

The thing that made this so hard for me is that I have no idea what a pytest mark does. I try not to use any of its magic or decorator features in favor of more explicit code. Thus when I got an error involving whatever a mark, I was quite confused.


At the very least, I hope posting this will help some poor soul who googles the error message, and instead of getting nothing like I did, maybe this will come up and then they will realize they forgot to `raise AttributeError` at the end of their `__getattr__`.

--- a/changelog/8265.improvement.rst
+++ b/changelog/8265.improvement.rst
@@ -1,0 +1,5 @@
+Improved error message when a module-level ``__getattr__`` fails to raise ``AttributeError`` for missing attributes.
+
+When a module defines a custom ``__getattr__`` that returns ``None`` instead of raising ``AttributeError``,
+pytest now provides a more helpful error message explaining the likely cause and suggesting the fix,
+rather than the cryptic "got None instead of Mark" message.

--- a/changelog/8265.improvement.rst
+++ b/changelog/8265.improvement.rst
@@ -1,5 +1,4 @@
-Improved error message when a module-level ``__getattr__`` fails to raise ``AttributeError`` for missing attributes.
+Emit a ``PytestCollectionWarning`` when a module-level ``__getattr__`` returns ``None`` for ``pytestmark`` instead of raising ``AttributeError``.
 
-When a module defines a custom ``__getattr__`` that returns ``None`` instead of raising ``AttributeError``,
-pytest now provides a more helpful error message explaining the likely cause and suggesting the fix,
-rather than the cryptic "got None instead of Mark" message.
+Previously this caused a cryptic ``TypeError: got None instead of Mark`` error.
+Now pytest issues a helpful warning and continues collecting the module normally.

--- a/src/_pytest/mark/structures.py
+++ b/src/_pytest/mark/structures.py
@@ -463,6 +463,15 @@ def normalize_mark_list(
     for mark in mark_list:
         mark_obj = getattr(mark, "mark", mark)
         if not isinstance(mark_obj, Mark):
+            # Provide a helpful error message for a common mistake:
+            # a module-level __getattr__ that doesn't raise AttributeError
+            if mark_obj is None:
+                raise TypeError(
+                    "got None instead of Mark - "
+                    "this is likely caused by a module-level __getattr__ "
+                    "that does not raise AttributeError for missing attributes. "
+                    "Make sure __getattr__ raises AttributeError when the attribute is not found."
+                )
             raise TypeError(f"got {mark_obj!r} instead of Mark")
         yield mark_obj
 

--- a/src/_pytest/mark/structures.py
+++ b/src/_pytest/mark/structures.py
@@ -30,6 +30,7 @@ from _pytest.deprecated import PARAMETRIZE_NON_COLLECTION_ITERABLE
 from _pytest.outcomes import fail
 from _pytest.raises import AbstractRaises
 from _pytest.scope import ScopeName
+from _pytest.warning_types import PytestCollectionWarning
 from _pytest.warning_types import PytestUnknownMarkWarning
 
 
@@ -443,7 +444,18 @@ def get_unpacked_marks(
                 mark_list.append(item)
     else:
         mark_attribute = getattr(obj, "pytestmark", [])
-        if isinstance(mark_attribute, list):
+        if mark_attribute is None:
+            warnings.warn(
+                "Module defines a `__getattr__` which returns None for "
+                "'pytestmark' instead of raising AttributeError. "
+                "Make sure `__getattr__` raises AttributeError for "
+                "attributes it does not provide. "
+                "See https://github.com/pytest-dev/pytest/issues/8265",
+                PytestCollectionWarning,
+                stacklevel=2,
+            )
+            mark_list = []
+        elif isinstance(mark_attribute, list):
             mark_list = mark_attribute
         else:
             mark_list = [mark_attribute]
@@ -463,15 +475,6 @@ def normalize_mark_list(
     for mark in mark_list:
         mark_obj = getattr(mark, "mark", mark)
         if not isinstance(mark_obj, Mark):
-            # Provide a helpful error message for a common mistake:
-            # a module-level __getattr__ that doesn't raise AttributeError
-            if mark_obj is None:
-                raise TypeError(
-                    "got None instead of Mark - "
-                    "this is likely caused by a module-level __getattr__ "
-                    "that does not raise AttributeError for missing attributes. "
-                    "Make sure __getattr__ raises AttributeError when the attribute is not found."
-                )
             raise TypeError(f"got {mark_obj!r} instead of Mark")
         yield mark_obj
 

--- a/testing/test_mark.py
+++ b/testing/test_mark.py
@@ -1350,7 +1350,7 @@ def test_module_getattr_without_attributeerror(pytester: Pytester) -> None:
     """
     Test that a helpful error message is provided when a module-level
     __getattr__ fails to raise AttributeError.
-    
+
     Regression test for https://github.com/pytest-dev/pytest/issues/8265
     """
     pytester.makepyfile(
@@ -1364,9 +1364,11 @@ def test_module_getattr_without_attributeerror(pytester: Pytester) -> None:
         """
     )
     result = pytester.runpytest()
-    result.stdout.fnmatch_lines([
-        "*TypeError*got None instead of Mark*",
-        "*module-level __getattr__*",
-        "*AttributeError*",
-    ])
+    result.stdout.fnmatch_lines(
+        [
+            "*TypeError*got None instead of Mark*",
+            "*module-level __getattr__*",
+            "*AttributeError*",
+        ]
+    )
     assert result.ret != 0

--- a/testing/test_mark.py
+++ b/testing/test_mark.py
@@ -1344,3 +1344,29 @@ def test_fixture_disallowed_between_marks() -> None:
         @pytest.mark.usefixtures("tmp_path")
         def foo():
             raise NotImplementedError()
+
+
+def test_module_getattr_without_attributeerror(pytester: Pytester) -> None:
+    """
+    Test that a helpful error message is provided when a module-level
+    __getattr__ fails to raise AttributeError.
+    
+    Regression test for https://github.com/pytest-dev/pytest/issues/8265
+    """
+    pytester.makepyfile(
+        """
+        def __getattr__(key):
+            # Bug: should raise AttributeError, but returns None
+            return None
+
+        def test_something():
+            assert True
+        """
+    )
+    result = pytester.runpytest()
+    result.stdout.fnmatch_lines([
+        "*TypeError*got None instead of Mark*",
+        "*module-level __getattr__*",
+        "*AttributeError*",
+    ])
+    assert result.ret != 0

--- a/testing/test_mark.py
+++ b/testing/test_mark.py
@@ -1348,8 +1348,8 @@ def test_fixture_disallowed_between_marks() -> None:
 
 def test_module_getattr_without_attributeerror(pytester: Pytester) -> None:
     """
-    Test that a helpful error message is provided when a module-level
-    __getattr__ fails to raise AttributeError.
+    Test that a helpful warning is emitted when a module-level
+    __getattr__ returns None instead of raising AttributeError.
 
     Regression test for https://github.com/pytest-dev/pytest/issues/8265
     """
@@ -1363,12 +1363,12 @@ def test_module_getattr_without_attributeerror(pytester: Pytester) -> None:
             assert True
         """
     )
-    result = pytester.runpytest()
+    result = pytester.runpytest("-W", "always::pytest.PytestCollectionWarning")
     result.stdout.fnmatch_lines(
         [
-            "*TypeError*got None instead of Mark*",
-            "*module-level __getattr__*",
-            "*AttributeError*",
+            "*PytestCollectionWarning*__getattr__*returns None*AttributeError*",
         ]
     )
-    assert result.ret != 0
+    # The module is buggy (__getattr__ returns None for all attributes),
+    # so no tests are collected, but pytest should NOT crash with a TypeError.
+    assert result.ret != ExitCode.INTERNAL_ERROR

